src/main/java/frc/robot/subsystems/driveTrain/features/SwerveModule.java

```diff
@@ -0,0 +1,156 @@
+package frc.robot.subsystems.driveTrain.features;
+
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.can.TalonFXConfiguration;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+import com.ctre.phoenix.sensors.AbsoluteSensorRange;
+import com.ctre.phoenix.sensors.CANCoderConfiguration;
+import com.ctre.phoenix.sensors.SensorInitializationStrategy;
+import com.ctre.phoenix.sensors.SensorTimeBase;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import pid.CtreMotionMagicController;
+import pid.CtrePIDController;
+import pid.PIDControlMode;
+import pid.interfaces.PIDController;
+import sensors.counter.CtreCANCoder;
+import sensors.counter.CtreEncoder;
+import sensors.counter.TalonEncoder;
+
+import static frc.robot.subsystems.driveTrain.DriveTrainConstants.*;
+
+
+public class SwerveModule {
+    private final WPI_TalonFX driveMotor;
+    private final WPI_TalonFX turningMotor;
+    private final CtreCANCoder absoluteEncoder;
+    private final CtreMotionMagicController turningController;
+    private final PIDController driveController;
+
+    public double getAngleOffset() {
+        return angleOffset;
+    }
+
+    private final double angleOffset;
+
+    public SwerveModule(int driverID, int turningId, int encoderID, double angleOffset) {
+        driveMotor = new WPI_TalonFX(driverID);
+        driveMotor.configAllSettings(swerveDriveFXConfig());
+
+        turningMotor = new WPI_TalonFX(turningId);
+        turningMotor.configAllSettings(swerveRotationFXConfig());
+
+        absoluteEncoder = new CtreCANCoder(encoderID);
+        absoluteEncoder.configAllSettings(swerveCANCoderConfig());
+
+        CtreEncoder turningEncoder = new TalonEncoder(turningMotor);
+        turningController = new CtreMotionMagicController(
+                turningMotor,
+                turningEncoder,
+                TURNING_MOTOR_PID,
+                ACC,
+                CRUISE_VEL,
+                ACC_SMOOTHING
+        );
+
+        CtreEncoder driverEncoder = new TalonEncoder(driveMotor);
+        driveController = new CtrePIDController(
+                driveMotor,
+                driverEncoder,
+                DRIVE_MOTOR_PID,
+                PIDControlMode.Velocity
+        );
+        this.angleOffset = angleOffset;
```

Comment:
space

Reply:
between what?

Reply:
d, i did space between end of constructor to offset because they are different yay